src/stan/variational/advi.hpp

```diff
@@ -0,0 +1,665 @@
+
+          // Check for convergence every "refresh_"th iteration
+          if (iter_counter % refresh_ == 0) {
+            elbo_prev = elbo;
+            elbo = calc_ELBO(musigmatilde);
+            delta_elbo = rel_decrease(elbo, elbo_prev);
+            cb.push_back(delta_elbo);
+            delta_elbo_ave = std::accumulate(cb.begin(), cb.end(), 0.0)
+                             / (double)(cb.size());
+            delta_elbo_med = circ_buff_median(cb);
+            *print_stream_
+                      << "  "
+                      << std::setw(4) << iter_counter
+                      << "  "
+                      << std::right << std::setw(9) << std::setprecision(1)
+                      << elbo
+                      << "  "
+                      << std::setw(16) << std::fixed << std::setprecision(3)
+                      << delta_elbo_ave
+                      << "  "
+                      << std::setw(15) << std::fixed << std::setprecision(3)
+                      << delta_elbo_med;
+
+            if (diag_stream_) {
+              end = clock();
+              delta_t = static_cast<double>(end - start) / CLOCKS_PER_SEC;
+
+              print_vector.clear();
+              print_vector.push_back(delta_t);
+              print_vector.push_back(elbo);
+              services::io::write_iteration_csv(*diag_stream_,
+                                                iter_counter, print_vector);
+            }
+
+            if (delta_elbo_ave < tol_rel_obj) {
+              *print_stream_ << "   MEAN ELBO CONVERGED";
+              do_more_iterations = false;
+            }
+
+            if (delta_elbo_med < tol_rel_obj) {
+              *print_stream_ << "   MEDIAN ELBO CONVERGED";
+              do_more_iterations = false;
+            }
+
+            if (iter_counter > 100) {
+              if (delta_elbo_med > 0.5 || delta_elbo_ave > 0.5) {
+                *print_stream_ << "   MAY BE DIVERGING... INSPECT ELBO";
+              }
+            }
+
+            *print_stream_ << std::endl;
+          }
+
+          // Check for max iterations
+          if (iter_counter == max_iterations) {
+            *print_stream_ << "MAX ITERATIONS REACHED" << std::endl;
+            do_more_iterations = false;
+          }
+
+          ++iter_counter;
+        }
+      }
+
+      void run_fullrank(double tol_rel_obj, int max_iterations) {
+        *print_stream_
+        << "This is stan::variational::advi::run_fullrank()" << std::endl;
+
+        // Initialize variational parameters: mu, L
+        Eigen::VectorXd mu = cont_params_;
+        Eigen::MatrixXd L  = Eigen::MatrixXd::Identity(model_.num_params_r(),
+                                                       model_.num_params_r());
+        advi_params_fullrank muL = advi_params_fullrank(mu, L);
+
+        // Robbins Monro ADAgrad
+        do_robbins_monro_adagrad(muL, tol_rel_obj, max_iterations);
+
+        cont_params_ = muL.mu();
+
+        // *print_stream_
```

Comment:
You should assume people reading the code know C++. 

In particular, you can assume they know mu = and L = ...
are assignments.

And you named the function "do_robbins_monro_adagrad", so you
don't need a comment that repeats the function name.

(In general, I'm opposed to function names that have "do" in
the title just to make them verbs, but I can live with it.)

Also, I'm not seeing the rest of the code, but is it really
necessary to create an identity matrix?  That's a quadratic
operation in the number of params and identity matrices aren't
usually good for much.

- Bob

> On Apr 29, 2015, at 3:32 AM, Rob Trangucci <notifications@github.com> wrote:
> 
> In src/stan/variational/advi.hpp:
> 
> > +      void run_fullrank(double tol_rel_obj, int max_iterations) {
> > +        *print_stream_
> > +        << "This is stan::variational::advi::run_fullrank()" << std::endl;
> > +
> > +        // Initialize variational parameters: mu, L
> > +        Eigen::VectorXd mu = cont_params_;
> > +        Eigen::MatrixXd L  = Eigen::MatrixXd::Identity(model_.num_params_r(),
> > +                                                       model_.num_params_r());
> > +        advi_params_fullrank muL = advi_params_fullrank(mu, L);
> > +
> > +        // Robbins Monro ADAgrad
> > +        do_robbins_monro_adagrad(muL, tol_rel_obj, max_iterations);
> > +
> > +        cont_params_ = muL.mu();
> > +
> > +        // *print_stream_
> 
> @akucukelbir Do you need these comments here?
> 
> —
> Reply to this email directly or view it on GitHub.
>